modules/system-disaster-recovery/src/test/java/org/apache/ignite/internal/disaster/system/SystemDisasterRecoveryManagerImplTest.java

```diff
@@ -0,0 +1,490 @@
+        prepareNodeStateForClusterReset();
+
+        when(messagingService.invoke(any(ClusterNode.class), any(), anyLong()))
+                .thenReturn(completedFuture(successResponseMessage));
+
+        CompletableFuture<Void> future = manager.resetCluster(List.of(thisNodeName, node2.name(), node3.name()));
+        assertThat(future, willCompleteSuccessfully());
+
+        verify(restarter).initiateRestart();
+    }
+
+    @Test
+    void resetClusterInitiatesRestartWhenMajorityOfCmgNodesRespondsWithOk() {
+        when(topologyService.allMembers()).thenReturn(List.of(thisNode, node2, node3, node4, node5));
+        prepareNodeStateForClusterReset();
+
+        respondSuccessfullyFrom(thisNode, node2);
+        respondWithExceptionFrom(node3, node4, node5);
+
+        CompletableFuture<Void> future = manager.resetCluster(List.of(thisNodeName, node2.name(), node3.name()));
+        assertThat(future, willCompleteSuccessfully());
+
+        verify(restarter).initiateRestart();
+    }
+
+    private void respondSuccessfullyFrom(ClusterNode... nodes) {
+        for (ClusterNode node : nodes) {
+            respondSuccessfullyFrom(node);
+        }
+    }
+
+    private void respondSuccessfullyFrom(ClusterNode node) {
+        when(messagingService.invoke(eq(node), any(), anyLong()))
+                .thenReturn(completedFuture(successResponseMessage));
+    }
+
+    private void respondWithExceptionFrom(ClusterNode... nodes) {
+        for (ClusterNode node : nodes) {
+            respondWithExceptionFrom(node);
+        }
+    }
+
+    private void respondWithExceptionFrom(ClusterNode node) {
+        when(messagingService.invoke(eq(node), any(), anyLong()))
+                .thenReturn(failedFuture(new TimeoutException()));
+    }
+
+    @Test
+    @DisplayName("resetCluster() fails and does not restart when majority of new CMG nodes do not respond")
+    void resetClusterFailsWhenNewCmgMajorityDoesNotRespond() {
+        when(topologyService.allMembers()).thenReturn(List.of(thisNode, node2, node3, node4, node5));
+        prepareNodeStateForClusterReset();
+
+        respondSuccessfullyFrom(thisNode, node4, node5);
+        respondWithExceptionFrom(node2, node3);
+
+        CompletableFuture<Void> future = manager.resetCluster(List.of(thisNodeName, node2.name(), node3.name()));
+        ClusterResetException ex = assertWillThrow(future, ClusterResetException.class, 10, SECONDS);
+        assertThat(ex.getMessage(), is("Did not get successful responses from new CMG majority, failing cluster reset."));
+
+        verify(restarter, never()).initiateRestart();
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {false, true})
+    void savesToVaultWhenGetsMessage(boolean fromSelf) throws Exception {
+        NetworkMessageHandler handler = extractMessageHandler();
+
+        ClusterNode conductor = fromSelf ? thisNode : node2;
+        handler.onReceived(resetClusterMessageOn2Nodes(conductor.name()), conductor, 0L);
```

Comment:
Can we actually send a message instead of doing this work-around job of extracting a handler and calling its methods? Looks too complicated.
Also, `extractMessageHandler` starts a manager, which is not evident from the name

Reply:
To actually send a message, we need a real implementation, a real `ClusterService`. This will bring another source of failures to the test, and the setup code is not that shorter.

As for `extractMessageHandler()`, the fact that it starts something is not important in this context. It actually stops the manager first. This is just a little plumbing to make it possible to test using a unit test without the real network (which is irrelevant to the logic under test).

Reply:
All I'm saying is that this code looks very fragile and very confusing. Apparently it assumes that you can call stop/start multiple times, which is not an invariant that we usually support. Is this even a helpful test?